### PR TITLE
Direct address configuration #297

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#297

Need to add tests.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs [hunk 1]
```diff
@@ -87,6 +86,21 @@ public long RequestQueueLimit
         /// </summary>
         public TimeSpan ShutdownTimeout { get; set; } = TimeSpan.FromSeconds(5);
 
+        internal List<string> ServerAddresses { get; private set; }
+
+        /// <summary>
+        /// Bind to given IP addresses and ports.
+        /// </summary>
+        public void Listen(params string[] addresses)
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: Do we want a different API here? Kestrel listens on `Endpoints` for example but I'm not sure how much more value that brings. Maybe strings are sufficient.

File: src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs [hunk 2]
```diff
@@ -71,7 +69,36 @@ public void Start<TContext>(IHttpApplication<TContext> application)
...
 
-            ParseAddresses(_serverAddresses.Addresses, Listener);
+            var serverAdressesPresent = _serverAddresses.Addresses.Count > 0;
+
+            if (_options.ServerAddresses?.Count > 0)
+            {
+                if (serverAdressesPresent)
+                {
+                    LogHelper.LogWarning(_logger, $"Overriding address(es) '{string.Join(", ", _serverAddresses.Addresses)}'. " +
+                        $"Binding to endpoints defined in {nameof(WebHostBuilderHttpSysExtensions.UseHttpSys)}() instead.");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs]: defined in HttpSysOptions

File: src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs [hunk 3]
```diff
@@ -87,6 +86,21 @@ public long RequestQueueLimit
         /// </summary>
         public TimeSpan ShutdownTimeout { get; set; } = TimeSpan.FromSeconds(5);
 
+        internal List<string> ServerAddresses { get; private set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: Redundant with UrlPrefixCollection

File: samples/SelfHostServer/Startup.cs [hunk 4]
```diff
@@ -37,6 +37,7 @@ public static void Main(string[] args)
                 .UseStartup<Startup>()
                 .UseHttpSys(options =>
                 {
+                    options.Listen("http://localhost:5000");
```
Comment (JunTaoLuo) [inline samples/SelfHostServer/Startup.cs]: To illustrate how to specify a specific address using the direct configuration api.

**Comment by JunTaoLuo** [comment]
:up::date:

File: src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs [hunk 5]
```diff
@@ -71,7 +69,37 @@ public void Start<TContext>(IHttpApplication<TContext> application)
...
+                    {
+                        _serverAddresses.Addresses.Add(prefix.FullPrefix);
+                    }
+                }
+            }
+            else if (serverAdressesPresent)
+            {
+                foreach (var value in _serverAddresses.Addresses)
+                {
+                    Listener.Options.UrlPrefixes.Add(UrlPrefix.Create(value));
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs]: You can use the simpler Add overload.

File: src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs [hunk 6]
```diff
@@ -80,12 +108,12 @@ public void Start<TContext>(IHttpApplication<TContext> application)
 
             _application = new ApplicationWrapper<TContext>(application);
 
-            if (_listener.Options.UrlPrefixes.Count == 0)
+            if (Listener.Options.UrlPrefixes.Count == 0)
             {
                 throw new InvalidOperationException("No address prefixes were defined.");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/MessagePump.cs]: This error should not be possible to hit anymore.

**Pull request merged**